http: allow ignoring unknown upgrade requests

ggreenway opened the pull request:
Fixes #36305

Add configuration to ignore either all unknown upgrade requests, or specific upgrade requests. This is allowed in https://datatracker.ietf.org/doc/html/rfc7230#section-6.7:

   A server MAY ignore a received Upgrade
   header field if it wishes to continue using the current protocol on
   that connection.

Risk Level: Medium
Testing: TODO
Docs Changes:
Release Notes: TODO

ggreenway commented:
Initial questions:

1) Is this the right way to ignore upgrades?
2) Is it better to default the global-ignore to true or false?
3) Do we need to handle h2/h3 specially? My understanding is that the Upgrade header doesn't apply to them at all (https://www.rfc-editor.org/rfc/rfc9113#name-the-upgrade-header-field).

alyssawilk commented:
cc @RyanTheOptimist for spec thoughts

alyssawilk commented on api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:
```diff
@@ -289,11 +289,18 @@ message HttpConnectionManager {
+    // If enabled, Envoy will ignore upgrade requests of this type. It is invalid to set both this option and
+    // :ref:`enabled <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.UpgradeConfig.enabled>`
+    // to true.
+    google.protobuf.BoolValue ignore = 4;
```
We can't retroactively make this a oneof for enabled / ignore right?

abeyad replied:
`oneof` is discouraged by the API style guide: https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md

```
Prefer multiple fields with defined precedence over boolean overloads of fields or oneof
...
This is more efficient on the wire. It also allows new alternatives to be added later in a way that allows control planes to be backward-compatible.
```

alyssawilk commented on api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:
```diff
@@ -289,11 +289,18 @@ message HttpConnectionManager {
+    // If enabled, Envoy will ignore upgrade requests of this type. It is invalid to set both this option and
+    // :ref:`enabled <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.UpgradeConfig.enabled>`
+    // to true.
+    google.protobuf.BoolValue ignore = 4;
```
actually follow-up on that, do we have or need a way to ignore for all types?

ggreenway replied:
What do you mean for "ignore for all types"? If you mean "ignore all upgrade requests", that's https://github.com/envoyproxy/envoy/pull/37150/files#diff-4b0f0f4bf391293548732ba0dac7ea417d196e9e5bb87ad4041850f453a982f5R802. If that's not what you meant, please clarify the question.

alyssawilk commented on envoy/http/filter_factory.h:
```diff
@@ -100,17 +100,24 @@ class FilterChainFactory {
   createFilterChain(FilterChainManager& manager, bool only_create_if_configured = false,
                     const FilterChainOptions& options = EmptyFilterChainOptions{}) const PURE;
 
+  enum class UpgradeAction {
```
I think move the comments below up here in case the enum gets used multiple places

alyssawilk reviewed (commented):
I'm fine with this conceptually but as commented would like @RyanTheOptimist (or @DavidSchinazi but he's less likely to respond) thoughts from a spec perspective
and tests, obviously but it's still a draft so can wait for those =P

RyanTheOptimist commented:
Yes, Upgrade is HTTP/1 only. However Extended CONNECT provides similar capabilities for H2 and H3. Do we have a similar configuration method for this?

chlunde commented on api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:
```diff
@@ -289,11 +289,18 @@ message HttpConnectionManager {
     // :ref:`upgrade documentation <arch_overview_upgrades>`.
     google.protobuf.BoolValue enabled = 3;
+
+    // If enabled, Envoy will ignore upgrade requests of this type. It is invalid to set both this option and
```
Explicitly describe what this means, ignore could mean pass through or the header could be discarded. Also, there are two headers involved, connection and upgrade. The user should understand the implication for both of them

ggreenway commented:
As mentioned by @chlunde above, we also need to remove the parts of the `Connection` header for upgrades. When looking at that, I found other code that removes h2c upgrades, and found that there is another portion for h2c that needs to be removed (`http2-settings`). Are there similar bits for other types of upgrades? And if there are, is there any risk if they are not removed?

DavidSchinazi commented:
There's an important distinction between HTTP versions here:
* Upgrade is used only in HTTP/1.1, it is documented in [Section 7.8 of RFC 9110](https://www.rfc-editor.org/rfc/rfc9110#section-7.8)
* Extended CONNECT is used only in HTTP/2 and HTTP/3, it is documented in [RFC 8441](https://www.rfc-editor.org/rfc/rfc8441) and [RFC 9220](https://www.rfc-editor.org/rfc/rfc9220)

(As a quick side note, RFC 7230 has been obsoleted by RFC 9110 and RFC 9112, please do not refer to it as some parts of it are incorrect).

They behave differently: in particular, in HTTP/1.1 the server can "ignore" the Upgrade and still send a successful HTTP/1.1 response without upgrading to a different protocol. In HTTP/2 and 3 however, that is not possible. If the server does not wish to use the other protocol, it has no choice but to reject the request.

This gets particularly tricky because Envoy normalizes every Extended CONNECT internally to look like Upgrade. So the information of whether ignoring is allowed or not gets somewhat hard to track down inside Envoy.

I worry that this will cause protocol confusion or request smuggling attacks, unless there are strong safeguards in place to ensure that this cannot happen in HTTP/2 or HTTP/3.

ggreenway commented:
@DavidSchinazi thanks for the clarification on this. I added a check, and renamed the config fields, to document and ensure that ignoring upgrades only happens on http 1.1.

DavidSchinazi commented:
Thanks @ggreenway ! As far as I can tell, this looks safe from a spec perspective. But I'll let folks more familiar with Envoy internals do the detailed code review.

ipoval commented:
@ggreenway 👋 , hi, what is going to be the expected behavior after this fix for the HTTP2 upstreams? Is it going to be safe to use HTTP2 clusters when clients supply -H "Connection: Upgrade" -H "Upgrade: TLS/1.3" headers for example? Just want to double check expectations after reading (https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-upgradeconfig) that upgrade config does not work with HTTP2

ggreenway commented:
@ipoval when Envoy ignores a downstream http/1.1 upgrade request, it will remove the upgrade headers from the request, so the request should be fine for any upstream http version because it no longer requests an upgrade.

ggreenway commented:
Integration tests reveal that this doesn't work. The http1 codec handles request processing differently if it thinks an upgrade is happening; see `handling_upgrade_` in http1/codec_impl.cc.

ggreenway commented:
A fix for this would probably have to be in the http1 codec_impl, but that doesn't have access to the configured upgrades which are processed later, so it would probably need to be a change that specifically ignores TLS upgrades.
